From 3960134e400a5d73fdfb04013ee1f7c3d8c77eeb Mon Sep 17 00:00:00 2001
From: Kartatz <105828205+Kartatz@users.noreply.github.com>
Date: Fri, 24 Oct 2025 20:51:19 -0300
Subject: [PATCH] Disable FDSan (File Descriptor Sanitizer) for the GCOBOL
 compiler on Android

Fixes the following issue:

$ aarch64-unknown-linux-android-gcobol main.cbl
fdsan: failed to exchange ownership of file descriptor: fd 5 is owned by FILE* 0x74acbbe9e0, was expected to be unowned
cobol1: internal compiler error: Aborted
---
 gcc/cobol/util.cc | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/gcc/cobol/util.cc b/gcc/cobol/util.cc
index d3a4b01a0..4765d8563 100644
--- a/gcc/cobol/util.cc
+++ b/gcc/cobol/util.cc
@@ -66,6 +66,10 @@
 #pragma GCC diagnostic ignored "-Wunused-result"
 #pragma GCC diagnostic ignored "-Wmissing-field-initializers"
 
+#if defined(__ANDROID__)
+extern "C" int android_fdsan_set_error_level(int) __attribute__((__weak__));
+#endif
+
 // External declarations.
 extern FILE * yyin;
 extern int yyparse(void);
@@ -2482,6 +2486,10 @@ cobol_parse_files (int nfile, const char **files)
     }
   }
   assert(os_locale.codeset);
+  
+  #if defined(__ANDROID__)
+  if (android_fdsan_set_error_level != NULL) android_fdsan_set_error_level(0);
+  #endif
 
   for (int i = 0; i < nfile; i++) {
     parse_file (files[i]);
-- 
2.50.1

